aten/src/ATen/native/cuda/SpectralOps.cu

```diff
@@ -0,0 +1,317 @@
+#include <numeric>
+#include <iostream>
+
+namespace at { namespace native {
+
+__forceinline__
+static bool is_pow_of_two(long long int  x) {
+  return (x & (x - 1)) == 0;
+}
+
+// In real-to-complex transform, cuFFT only fills half of the values due to
+// conjugate symmetry. See native/SpectralUtils.h for more details.
+// The following structs are used to fill in the other half with symmetry in
+// case of real-to-complex transform with onesided=False flag.
+
+// counting_iterator => index to fill
+struct cnt_to_dst_idx_functor : public thrust::unary_function<int64_t, int64_t>
+{
+  const int64_t last_dim_size;
+  const int64_t last_dim_start_slice;
+  const int64_t last_dim_to_fill_size;
+
+  cnt_to_dst_idx_functor(int64_t last_dim_size, int64_t last_dim_start_slice) :
+    last_dim_size(last_dim_size), last_dim_start_slice(last_dim_start_slice),
+    last_dim_to_fill_size(last_dim_size - last_dim_start_slice) {}
+
+  __host__ __device__ __forceinline__
+  int64_t operator()(const int64_t& i) const
+  {
+    int64_t imag = i % 2;
+    int64_t idx = i / 2;
+    int64_t num_dim = idx / last_dim_to_fill_size;
+    int64_t slice_idx = idx % last_dim_to_fill_size;
+    return (num_dim * last_dim_size + last_dim_start_slice + slice_idx) * 2 + imag;
+  }
+};
+
+// index to fill => index to read from
+template <typename scalar_t>
+struct dst_idx_to_src_functor : public thrust::unary_function<int64_t, scalar_t>
+{
+  // output can have at most dim 5 (batch + 3 signal dim + real/imag)
+  int64_t sizes[5], strides[5];
+  const int64_t signal_ndim;
+  scalar_t *data;  // device ptr
+
+  dst_idx_to_src_functor(const Tensor& batched_complex_signal)
+    : signal_ndim(batched_complex_signal.dim() - 1),
+      data(batched_complex_signal.data<scalar_t>()) {
+    for (int64_t i = 0; i < signal_ndim; i++) {
+      sizes[i] = batched_complex_signal.size(i);
+      strides[i] = batched_complex_signal.stride(i);
+    }
+  }
+
+  __device__ __forceinline__
+  scalar_t operator()(const int64_t& write_idx_with_imag) const
+  {
+    int64_t imag = write_idx_with_imag % 2;
+    // all but first (batch) and last (real/imag) dims need to be reflected
+    int64_t read_idx = 0;
+    int64_t remainder = write_idx_with_imag - imag;
+    int64_t dim_idx, dim_stride;
+    for (int64_t i = 0; i < signal_ndim; i++) {
+      dim_stride = strides[i];
+      dim_idx = remainder / dim_stride;
+      if (i == 0) {
+        read_idx += dim_idx * dim_stride;
+      } else if (dim_idx != 0) {
+        read_idx += (sizes[i] - dim_idx) * dim_stride;
+      }
+      remainder = remainder % dim_stride;
+    }
+    if (imag) {
+      return -data[read_idx + 1];
+    } else {
+      return data[read_idx];
+    }
+  }
+};
+
+// input should be a contiguous batched tensor of same size as full (twosided)
+// signals, but only contains half (onesided) of the values.
+// This function modifies inplace.
+__forceinline__
+static void _fft_fill_with_conjugate_symmetry_(Tensor& input,
```

Comment:
This thrust CUDA implementation looks suspiciously low level. Do we really not have something that already does this for us? This is just reflecting half of a tensor buffer, is that right?

Reply:
It is more tricky than that. For a signal of sizes `N_1, ..., N_k`. We fill the each value in `tensor[:, :, ..., :, N_k / 2:)` by `tensor[x_1, ..., x_k] = tensor[y_1, ..., y_k]*`, where `*` is the conjugate op, and `y_i = (N_i - x_i) mod N_i`.